### ISSUE TITLE
ORM toIterable broken due to ForwardCompatibilty result

### ISSUE THREAD

**Issue by bobvandevijver** [doctrine/dbal#4588]
The following exception occurs when using `toIterable` on a query generated by a repository query builder:

```
Argument 1 passed to Doctrine\ORM\Internal\Hydration\AbstractHydrator::toIterable() must implement interface Doctrine\DBAL\Driver\Statement, instance of Doctrine\DBAL\ForwardCompatibility\Result given, called in /var/www/some-project/vendor/doctrine/orm/lib/Doctrine/ORM/AbstractQuery.php on line 1010
```

The code line triggering this exception is as follows:
```
$userIterator = $this->userRepository->createQueryBuilder('u')->getQuery()->toIterable();
```

This started after upgrading this library from 1.12.1 to 1.13.0. It looks like the following line is the cause of this:

https://github.com/doctrine/dbal/blob/79a6a843d49de8676efd38f30d29023c69ebff02/lib/Doctrine/DBAL/Connection.php#L1313

Which triggers the exception finally in ORM in the [`AbstractQuery`](https://github.com/doctrine/orm/blob/a588555ecd837b8d7e89355d9a13902e54d529c7/lib/Doctrine/ORM/AbstractQuery.php#L1010), where the `toIterable` doesn't receive the statement object it expects.

For me it looks like the `ForwardCompatibilty` isn't fully compatible with the statement object returned previously, so it would seem to be it needs to be fixed here? If that is not the case and this issue needs to be opened on the ORM repository, just say the word!

**Comment by TwanVermeulen** [comment]
I get a similar exception after upgrading to 1.13.0. But in a different vendor:

`Argument 1 passed to Symfony\Component\Security\Acl\Dbal\AclProvider::hydrateObjectIdentities() must implement interface Doctrine\DBAL\Driver\Statement, instance of Doctrine\DBAL\ForwardCompatibility\Result given, called in /var/www/app/default/symfony-standard/app_manager/vendor/symfony/security-acl/Dbal/AclProvider.php on line 468`

**Comment by mdumoulin** [comment]
@bobvandevijver Seems the issue is rather on ORM side. The AbstractHydrator::toIterable() method should rely on Doctrine\DBAL\Driver\ResultStatement instead of Doctrine\DBAL\Driver\Statement.

If I'm correct, this issue was already resolved and you can fix it by updating the ORM version : https://github.com/doctrine/orm/commit/930859f803ae1582bbdd0614b0e26def5d663b54#diff-f1936bc3177c94ebd06906a54a439e496b410cfee881d6740ac344f03382e0dd

**Comment by bobvandevijver** [comment]
@mdumoulin You're right, it is indeed fixed in ORM 2.8.3. 👍🏻

@TwanVermeulen Looks like there is an PR pending to fix your issue: https://github.com/symfony/security-acl/pull/64
